src/meds_torch/models/components/transformer_decoder.py

```diff
@@ -1,15 +1,286 @@
+        mask: Optional attention mask
+        is_time_budget: Whether using time-based budget
+        min_time_len: Minimum time length for generation
+        max_seq_len: Maximum sequence length to generate
+
+    Returns:
+        Tuple containing:
+        - Updated cumulative time tensor
+        - Boolean indicating whether to continue generation
+
+    Examples:
+        >>> import torch
+        >>> B = 2  # batch_size
+        >>> device = 'cpu'
+        >>> # Test time budget
+        >>> cumulative = torch.zeros(B, device=device)
+        >>> # Create a single token mask
+        >>> current_sample = torch.randint(0, 5, (5,), device=device)
+        >>> get_time = lambda x: torch.ones(B, device=device)
+        >>> time_budget = GenerationBudget.from_time_len(1.5)  # Generate until cumulative time > 1.5
+        >>> new_time, continue_gen = update_generation_budget(
+        ...     cumulative_time=cumulative,
+        ...     current_sample=current_sample,
+        ...     num_generated_tokens=2,
+        ...     get_next_token_time=get_time,
+        ...     budget=time_budget,
+        ... )
+        >>> new_time  # Should be 1
+        tensor([1., 1.])
+        >>> continue_gen  # Should continue since < min_time_len
+        True
+        >>> new_time, continue_gen = update_generation_budget(
+        ...     cumulative_time=new_time,
+        ...     current_sample=current_sample,
+        ...     num_generated_tokens=3,
+        ...     get_next_token_time=get_time,
+        ...     budget=time_budget,
+        ... )
+        >>> continue_gen
+        False
+        >>> new_time # Should be incremented by 1
+        tensor([2., 2.])
+
+        >>> seq_len_budget = GenerationBudget.from_seq_len(10)  # Generate until max_seq_len >= 10
+        >>> # Test sequence length budget
+        >>> _, continue_gen = update_generation_budget(
+        ...     cumulative_time=cumulative,
+        ...     current_sample=current_sample,
+        ...     num_generated_tokens=5,
+        ...     get_next_token_time=None,
+        ...     budget=seq_len_budget,
+        ... )
+        >>> continue_gen  # Should continue
+        True
+        >>> seq_len_budget = GenerationBudget.from_seq_len(5)  # Generate until max_seq_len >= 10
+        >>> # Test sequence length budget
+        >>> _, continue_gen = update_generation_budget(
+        ...     cumulative_time=cumulative,
+        ...     current_sample=current_sample,
+        ...     num_generated_tokens=5,
+        ...     get_next_token_time=None,
+        ...     budget=seq_len_budget,
+        ... )
+        >>> continue_gen # Should stop since > max_seq_len
+        False
+    """
+    # Get budget constraints
+    min_time_len = budget.min_time_len
+    is_time_budget = exists(min_time_len)
+    if is_time_budget:
+        # Get time prediction for the new token
+        with torch.no_grad():
+            if get_next_token_time is None:
```

Comment:
I would pull this error case out into the top to simplify your flow -- e.g., `if is_time_budget and get_next_token_time is None: ...` and also test it in your doctests.